src/t_hash.c

```diff
@@ -689,49 +751,124 @@ void hashTypeRandomElement(robj *hashobj, unsigned long hashsize, listpackEntry
+/* 'argc_start' parameter position of the first field.
+ * About the flags and in_flags, please see hashTypeSet. */
+void hsetGenericCommand(client *c, int argc_start, int flags, int in_flags) {
     robj *o;
-    if ((o = hashTypeLookupWriteOrCreate(c,c->argv[1])) == NULL) return;
+    robj *key = c->argv[1];
 
-    if (hashTypeExists(o, c->argv[2]->ptr)) {
-        addReply(c, shared.czero);
-    } else {
-        hashTypeTryConversion(o,c->argv,2,3);
-        hashTypeSet(o,c->argv[2]->ptr,c->argv[3]->ptr,HASH_SET_COPY);
-        addReply(c, shared.cone);
-        signalModifiedKey(c,c->db,c->argv[1]);
-        notifyKeyspaceEvent(NOTIFY_HASH,"hset",c->argv[1],c->db->id);
-        server.dirty++;
+    int added = 0;    /* Number of new element added. */
+    int updated = 0;  /* Number of elements with updated value. */
+
+    /* We expect to have an even number of (argc - argc_start).
+     * Since we expect any number of field-value pairs here. */
+    int elements = c->argc - argc_start;
+    if (elements % 2 || !elements) {
+        addReplyErrorFormat(c, "wrong number of arguments for '%s' command",
+                            c->cmd->name);
+        return;
     }
-}
 
-void hsetCommand(client *c) {
-    int i, created = 0;
-    robj *o;
+    /* Turn options into simple to check vars. */
+    int nx = (in_flags & HSET_IN_NX) != 0;
+    int xx = (in_flags & HSET_IN_XX) != 0;
+    int ch = (in_flags & HSET_IN_CH) != 0;
 
-    if ((c->argc % 2) == 1) {
-        addReplyErrorFormat(c,"wrong number of arguments for '%s' command",c->cmd->name);
+    /* Check for incompatible options. */
+    if (nx && xx) {
+        addReplyError(c, "XX and NX options at the same time are not compatible");
         return;
     }
 
-    if ((o = hashTypeLookupWriteOrCreate(c,c->argv[1])) == NULL) return;
-    hashTypeTryConversion(o,c->argv,2,c->argc-1);
+    /* Lookup the key and check the type. */
+    o = lookupKeyWrite(c->db, key);
+    if (checkType(c, o, OBJ_HASH)) return;
+
+    if (o == NULL) {
+        /* No key + XX option: nothing to do. */
+        if (xx) {
+            addReply(c, shared.czero);
+            return;
+        }
+
+        /* Create the hash object. */
+        o = createHashObject();
+        dbAdd(c->db, key, o);
+    }
+    hashTypeTryConversion(o, c->argv, argc_start, c->argc-1);
+
+    /* Process all elements. */
+    for (int i = argc_start; i < c->argc; i += 2) {
+        int retflags = 0;
+
+        hashTypeSet(o, c->argv[i]->ptr, c->argv[i+1]->ptr, flags, in_flags, &retflags);
+
+        if (retflags & HSET_OUT_ADDED) added++;
+        if (retflags & HSET_OUT_UPDATED) updated++;
+    }
 
-    for (i = 2; i < c->argc; i += 2)
-        created += !hashTypeSet(o,c->argv[i]->ptr,c->argv[i+1]->ptr,HASH_SET_COPY);
+    int dirty = added + updated;
 
     /* HMSET (deprecated) and HSET return value is different. */
     char *cmdname = c->argv[0]->ptr;
     if (cmdname[1] == 's' || cmdname[1] == 'S') {
```

Comment:
i think at this point (calling this function from a 3rd place), we should stop using this trick, and pass an argument)
or alternatively, since HMSET was the odd case, maybe do a full string compare to `HMSET` rather than look at the second letter.